### PR TITLE
Semantic text - field mapper

### PR DESCRIPTION
**Pull request by carlosdelest**
This creates a new `semantic_text` field mapper, that receives a `model_id` parameter. 

`model_id` will be used to perform inference on document ingestion and query time.

```json
PUT test-index
{
    "mappings": {
        "properties": {
            "my-inference-field": {
                "type": "semantic_text",
                "model_id": "my-model-id"
            }
      }
}

```

This field mapper does not index values - that will be taken care of by a metadata field mapper, as the inference results will be included in a different top level field.

A new `getInferenceModel()` method has been added to `MappedFieldType` so we'll be able to get information of the models used in the different fields from the `MappingLookup` and related structures in future PRs.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java [hunk 1]
```diff
@@ -203,6 +203,14 @@ public List<String> dimensions() {
         return Collections.emptyList();
     }
 
+    public String getInferenceModel() {
```
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:206]: Method added for future retrieval from `MappingLookup`

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:206]: @carlosdelest I would prefer a new interface & `instanceof` to check that something satisfies this new API. I don't like the idea of having it at the base MappedFieldType.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/LocalStateCompositeXPackPlugin.java [hunk 2]
```diff
@@ -447,7 +447,10 @@ public void onIndexModule(IndexModule indexModule) {
 
     @Override
     public Function<String, Predicate<String>> getFieldFilter() {
-        List<Function<String, Predicate<String>>> items = filterPlugins(MapperPlugin.class).stream().map(p -> p.getFieldFilter()).toList();
+        List<Function<String, Predicate<String>>> items = filterPlugins(MapperPlugin.class).stream()
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/LocalStateCompositeXPackPlugin.java:450]: This doesn't make sense to me. We shouldn't change this. Why do the other plugins that satisfy the mapping plugin don't suffer from this?

What is the actual issue here?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextFieldMapper.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        return (SemanticTextFieldMapper) in;
+    }
+
+    private static Builder builder(FieldMapper in) {
+        return ((SemanticTextFieldMapper) in).builder;
+    }
+
+    public static class Builder extends FieldMapper.Builder {
+
+        private final Parameter<String> modelId = Parameter.stringParam("model_id", false, m -> builder(m).modelId.get(), null)
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextFieldMapper.java:48]:
```suggestion
        private final Parameter<String> modelId = Parameter.stringParam("model_id", false, m -> toType(m).fieldType().modelId, null)
```

Or at least don't create a builder here. That seems unnecessary.

File: server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:206]: I am unclear on where the method would be called / where would the instanceof check go. Could you clarify?

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:206]: I agree on the sentiment, I just could not see what the impact is of an instanceof because the method is currently only exposed but never used. It is also true that this kind of special cases complicate the code in the long run, but I see what you mean.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:206]: @javanna I think @carlosdelest is trying to lay the ground work for gathering field capabilities in some way to determine if a model is used and what model. 

I agree, it's way too early here to make any changes to wide expansive interfaces like this one.

Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:206]: As Ben said, this will be used to gather the fields that are using an ML model so they can be included in `IndexMetadata` in a later stage.

I'm OK with using a new interface and defer the decision to include this method as part of the base class. 

@Mikep86, please be aware that I'll be moving `getInferenceModel()` from `MappedFieldType` to a new inteface.

Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java]: Addressed in 048779b7830843a43954e96b3ddc832b0cfb1ddd

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextFieldMapper.java (hunk 3 above)
Comment (carlosdelest) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextFieldMapper.java]: Good call - removing it

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/LocalStateCompositeXPackPlugin.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/LocalStateCompositeXPackPlugin.java:451]: Thanks for the heads up Ben!

Not doing this made some tests fail (see [this test result](https://gradle-enterprise.elastic.co/s/gflru57u7wcio/tests/task/:x-pack:plugin:ml:internalClusterTest/details/org.elasticsearch.license.MachineLearningLicensingIT/testAutoCloseJobWithDatafeed?top-execution=1)).

My understanding was that we were adding another MapperPlugin that was not accounted for - thus the change to understand that the MapperPlugin did something useful.

I've been checking and after merging with main this is no longer an issue, so I've withdrawn the changes.

**Review by original-brownbear — COMMENTED** [review]
Added a couple of smaller points on cleaning up the code a little :)

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextFieldMapper.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                modelId.getValue(),
+                copyTo,
+                this
+            );
+        }
+    }
+
+    public static final TypeParser PARSER = new TypeParser((n, c) -> new Builder(n), notInMultiFields(CONTENT_TYPE));
+
+    public static class SemanticTextFieldType extends SimpleMappedFieldType implements InferenceModelFieldType {
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextFieldMapper.java:96]: Makes sense - done in 4b066556447b00c9fbd6b2995b5c5b8a961b7c70

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextFieldMapper.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            return SourceValueFetcher.toString(name(), context, format);
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(FieldDataContext fieldDataContext) {
+            throw new IllegalArgumentException("[semantic_text] fields do not support sorting, scripting or aggregating");
+        }
+    }
+
+    private final String modelId;
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextFieldMapper.java]: Good call - we could always retrieve it from the field type. Addressed in 39e061fc9fc9e6e82712d81629081c4f49e885cb

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextFieldMapper.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    private final Builder builder;
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextFieldMapper.java]: Thanks for the context. I removed the need for storing the Builder previously, so I removed builder refs entirely in 4b066556447b00c9fbd6b2995b5c5b8a961b7c70

File: test/test-clusters/src/main/java/org/elasticsearch/test/cluster/FeatureFlag.java [hunk 7]
```diff
@@ -16,7 +16,7 @@
  */
 public enum FeatureFlag {
     TIME_SERIES_MODE("es.index_mode_feature_flag_registered=true", Version.fromString("8.0.0"), null),
-    LEARNING_TO_RANK("es.learning_to_rank_feature_flag_enabled=true", Version.fromString("8.12.0"), null),
+    LEARN_TO_RANK("es.learn_to_rank_feature_flag_enabled=true", Version.fromString("8.10.0"), null),
```
Comment (benwtrent) [inline test/test-clusters/src/main/java/org/elasticsearch/test/cluster/FeatureFlag.java:19]: seems like an unnecessary change? Merge in main?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/LocalStateCompositeXPackPlugin.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/LocalStateCompositeXPackPlugin.java:451]: OK, but I still see the change here, you are going to revert this in a future commit here?

File: x-pack/plugin/src/yamlRestTest/java/org/elasticsearch/xpack/test/rest/XPackRestIT.java [hunk 8]
```diff
@@ -43,7 +43,7 @@ public class XPackRestIT extends AbstractXPackRestTest {
         .setting("xpack.searchable.snapshot.shared_cache.region_size", "256KB")
         .user("x_pack_rest_user", "x-pack-test-password")
         .feature(FeatureFlag.TIME_SERIES_MODE)
-        .feature(FeatureFlag.LEARNING_TO_RANK)
+        .feature(FeatureFlag.LEARN_TO_RANK)
```
Comment (benwtrent) [inline x-pack/plugin/src/yamlRestTest/java/org/elasticsearch/xpack/test/rest/XPackRestIT.java:46]: again, seems unnecessary? Merge in main?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/LocalStateCompositeXPackPlugin.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/LocalStateCompositeXPackPlugin.java]: Same case as the other one - fixed in d7d8d5e07f74dd7eebbb9567b2cb5f9400094e7d

File: test/test-clusters/src/main/java/org/elasticsearch/test/cluster/FeatureFlag.java (hunk 7 above)
Comment (carlosdelest) [reply test/test-clusters/src/main/java/org/elasticsearch/test/cluster/FeatureFlag.java]: Fixed in 07613ae8876d9dc76ee60faf270d7dd534e324e9

File: x-pack/plugin/src/yamlRestTest/java/org/elasticsearch/xpack/test/rest/XPackRestIT.java (hunk 8 above)
Comment (carlosdelest) [reply x-pack/plugin/src/yamlRestTest/java/org/elasticsearch/xpack/test/rest/XPackRestIT.java]: Same as before - fixed in 07613ae8876d9dc76ee60faf270d7dd534e324e9

**Review by benwtrent — APPROVED** [review]

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/LocalStateCompositeXPackPlugin.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/LocalStateCompositeXPackPlugin.java:451]: OK - removing this caused tests to fail. I've looked deeper into this issue:
- `LocalStateCompositeXPackPlugin` is used for testing
- No other subclasses of `LocalStateCompositeXPackPlugin` use a `MapperPlugin` as part of its plugins
- `LocalStateMachineLearning` has as plugins `Security` (which is a `MapperPlugin`) and `MachineLearning`. Making `MachineLearning` another `MapperPlugin` implied that there are 2 `MapperPlugins`
- The existing code does not check whether the actual `getFieldsFilter()` method has been overriden or not, but just the numbrer of `MapperPlugins`

So I think the changes I've done actually preserve the original intention - we're **not** looking for how many `MapperPlugins` there are, but how many of them actually override the method (that is, do not return `NOOP_FIELD_FILTER`.

Does it make sense? Happy to discuss if not

[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/LocalStateCompositeXPackPlugin.java:451]: @carlosdelest I see now. Seems good to me

**Review by original-brownbear — APPROVED** [review]
